crates/ra_fmt2/src/engine.rs

```diff
@@ -37,33 +49,48 @@ impl FmtDiff {
     /// 
     /// # Arguments
     ///
-    /// * `block` - A `Block` that is always a token because rules match tokens.
+    /// * `left_blk` - A `Block` that is always a token, previous token used to determine
+    ///     space after token.
+    /// * `right_blk` - A `Block` that is always a token, check right preceding whitespace.
     /// * `rule` - A `SpaceRule`.
-    fn check_spacing(&self, rule: &SpacingRule, block: &Block) {
-        let whitespace = block.get_whitespace();
-        if *whitespace.borrow() != *rule {
-            block.set_spacing(rule)
+    fn compute_spacing(
```

Comment:
To make `compute_spacing` more like your example it could return a SpaceBlock 
```rust
pub(crate) struct SpaceBlock {
    spaces: u32,
    newline: bool,
    indent: bool,
}
```
which `set_spacing` would take instead of a rule and just apply the `SpaceBlock` to the `Whitespace`